src/async_query.rs

```diff
@@ -0,0 +1,147 @@
+//! # Async query support
+//! This module does not export anything, as it provides additionnal 
+//! methods on [`WMIConnection`](WMIConnection)
+//!
+//! You only have to activate the `async-query` feature flag in Cargo.toml to use them.
+//! ```toml
+//! wmi = { version = "x.y.z",  features = ["async-query"] }
+//! ```
+//! 
+//! 
+
+use crate::BStr;
+use crate::result_enumerator::IWbemClassWrapper;
+use crate::{
+    connection::WMIConnection, utils::check_hres, WMIError,
+};
+use crate::query_sink::QuerySink;
+use std::ptr;
+use winapi::{
+    um::{
+        wbemcli::IWbemObjectSink,
+        wbemcli::{
+            WBEM_FLAG_BIDIRECTIONAL,
+        },
+    },
+};
+use wio::com::ComPtr;
+use serde::de;
+use futures::stream::{Stream, TryStreamExt, StreamExt};
+
+///
+/// ### Aditionnal async methods
+/// **Following methods are implemented under the
+/// `async-query` feature flag.**
+///
+impl WMIConnection {
+    /// Wrapper for the [ExecQueryAsync](https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemservices-execqueryasync)
+    /// method. Provides safety checks, and returns results
+    /// as a Stream instead of the original Sink.
+    ///
+    pub fn exec_query_async_native_wrapper(
+        &self,
+        query: impl AsRef<str>,
+    ) -> Result<impl Stream<Item=Result<IWbemClassWrapper, WMIError>>, WMIError> {
+        let query_language = BStr::from_str("WQL")?;
+        let query = BStr::from_str(query.as_ref())?;
+
+        let (tx, rx) = async_channel::unbounded();
+        let p_sink: ComPtr<IWbemObjectSink> = QuerySink::new(tx);
+
+        unsafe {
+            check_hres((*self.svc()).ExecQueryAsync(
+                query_language.as_bstr(),
+                query.as_bstr(),
+                WBEM_FLAG_BIDIRECTIONAL as i32,
+                ptr::null_mut(),
+                p_sink.as_raw(),
+            ))?;   
+        }
+        Ok(rx)
+    }
+
+    /// Async version of [`raw_query`](WMIConnection#method.raw_query)
+    /// Execute a free-text query and deserialize the results.
+    /// Can be used either with a struct (like `query` and `filtered_query`),
+    /// but also with a generic map.
+    ///
+    /// ```edition2018
+    /// # fn main() -> Result<(), wmi::WMIError> {
+    /// # use wmi::*;
+    /// # use std::collections::HashMap;
+    /// # use futures::executor::block_on;
+    /// # let con = WMIConnection::new(COMLibrary::new()?.into())?;
+    /// let results: Vec<HashMap<String, Variant>> = block_on(con.async_raw_query("SELECT Name FROM Win32_OperatingSystem"))?;
+    /// #   Ok(())
+    /// # }
+    /// ```
+    pub async fn async_raw_query<T>(&self, query: impl AsRef<str>) -> Result<Vec<T>, WMIError>
```

Comment:
I did it this way for 2 reasons:
1. The API is consistent between this function and `pub fn raw_query<T>(&self, query: impl AsRef<str>) -> Result<Vec<T>, WMIError>`
2. It provides an `async` function, rather than a function that returns a future, which seems a bit more idiomatic

Can you confirm you want this instead? I'll do it your way ;)

Reply:
~It a shame to loose the async function syntax, but it's the correct abstraction over the actual API.
I would say that the other API should also have been an iterator, but for async users it's probably even more important.~

nah, you are right. A vec is nice and simple 👍

Reply:
Great! And there's always the `exec_async_query_native_wrapper` for those who wants a Stream ;)